### PR TITLE
Setting to use the advanced setting editor for the settings

### PR DESCRIPTION
**Pull request by echarles**
## References

Fixes https://github.com/jupyterlab/jupyterlab/issues/12283

## Code changes

Add a setting `useJsonEditor` in the `settingeditor-extension` package. If that setting is `true, the advanced json editor is use to edit the settings.

## User-facing changes

Untitled4

## Backwards-incompatible changes

None.

### CODE REVIEW

**Review by fcollonval — CHANGES_REQUESTED** [review]
I would use an enumeration instead of a boolean as setting; something like

Default settings editor: ['UI', 'json']

**Comment by jtpio** [comment]
Also, is this new option needed? Or could the menu customization via the settings be used to replace the menu entry by `"command": "settingeditor:open-json"`?

**Comment by echarles** [comment]
> I would use an enumeration instead of a boolean as setting; something like Default settings editor: ['UI', 'json']

Thx @fcollonval. Yes, we can do that.

> Also, is this new option needed? Or could the menu customization via the settings be used to replace the menu entry by "command": "settingeditor:open-json"?

Thx @jtpio. With  `option`, are you meaning the new setting, or the new command? I guess you mean the latter. I have thought replacing the menu entry (which is the visual menu show here below) based on the user setting ediotr (ui or json), but I wonder if it is possible to achieve the replace without the full browser refresh. Once the menu is built, is it mutable?

Screenshot 2022-04-26 at 10 23 48

**Comment by fcollonval** [comment]
I like a lot the idea of @jtpio - but I think it is nice to still combine it with a boolean settings as menu customization is not easy for the end users.

So to layout the changes needed, you would not add a new command, but extend the `open` one to receive arguments.

If the argument defines the type of settings editor, use the argument otherwise fallback to the settings value. And I think that for 4.0 it will make sense to drop the command `'settingeditor:open-json'`. But when backporting this enhancement we should keep it around.

**Comment by echarles** [comment]
> So to layout the changes needed, you would not add a new command, but extend the open one to receive arguments. If the argument defines the type of settings editor, use the argument otherwise fallback to the settings value. And I think that for 4.0 it will make sense to drop the command 'settingeditor:open-json'. But when backporting this enhancement we should keep it around.

@fcollonval I have implemented your suggestion in my latest commit. Maybe you could have a look and confirm this was the intent. If confirmed, I will then implement your suggestion `I would use an enumeration instead of a boolean as setting; something like Default settings editor: ['UI', 'json']`.

**Comment by echarles** [comment]
> and I think that for 4.0 it will make sense to drop the command 'settingeditor:open-json'

I have not been able to achieve that as the execution of the `open` and `open-json` commands are defined in separated plugins.

File: packages/settingeditor-extension/src/index.ts [hunk 1]
```diff
@@ -101,64 +101,77 @@ function activate(
...
+      });
     },
     label: trans.__('Settings Editor')
   });
 
+  if (palette) {
+    palette.addItem({
+      category: trans.__('Settings'),
+      command: CommandIDs.open,
+      args: { useJsonEditor: true }
```
Comment (fcollonval) [inline packages/settingeditor-extension/src/index.ts:171]:
```suggestion
      args: { useJsonEditor: false }
```

File: packages/settingeditor-extension/src/index.ts [hunk 2]
```diff
@@ -101,64 +101,77 @@ function activate(
...
-      void tracker.add(editor);
-      shell.add(editor);
+  commands.addCommand(CommandIDs.open, {
+    execute: async ({ useJsonEditor }: { useJsonEditor: boolean }) => {
+      registry.load(plugin.id).then(settings => {
+        if (useJsonEditor) {
+          commands.execute(CommandIDs.openJSON);
+          return;
+        }
+        (settings.get('useJsonEditor').composite as boolean)
```
Comment (fcollonval) [inline packages/settingeditor-extension/src/index.ts:159]: That's the spirit

```suggestion
    execute: async ({ useJsonEditor }: { useJsonEditor?: boolean }) => {
      registry.load(plugin.id).then(settings => {
        useJsonEditor ?? (settings.get('useJsonEditor').composite as boolean)
```

**Review by fcollonval — COMMENTED** [review]
Thanks @echarles  yes this is the idea.

**Comment by echarles** [comment]
@fcollonval I have merged with latest master and resolved the conflicts.

**Comment by echarles** [comment]
Galata is failing on `Open the settings editor with a specific search query `

I am not familiar with the recent setting search query. I run `openUi({ query: '' });` What should I do?

**Comment by echarles** [comment]
Did a rebase on latest master and UI test is now green. A few other flows are failing mostly on splice feature but should not berelated to this PR.

**Comment by echarles** [comment]
The UI regression test fails on ` Extension Manager › Search` which is not related to this PR.

@fcollonval I think this is ready to be merged.

**Comment by jtpio** [comment]
> The UI regression test fails on ` Extension Manager › Search` which is not related to this PR.

Yes also noticed in other places: https://github.com/jupyterlab/jupyterlab/issues/12340#issuecomment-1115833463

**Comment by echarles** [comment]
Just solved recent conflicts with master and CI is still green (exception of the not-revelant Extension Manager › Search).

**Comment by echarles** [comment]
@fcollonval @jtpio I have rebased and solved the conflicts. Happy to get your reviews so we can merge and backport.

**Review by fcollonval — APPROVED** [review]
Thanks @echarles LGTM

> CI failure is not related

**Comment by echarles** [comment]
@meeseeksdev please backport to 3.4.x

**Comment by echarles** [comment]
Thx @fcollonval 

The backport PR is https://github.com/jupyterlab/jupyterlab/pull/12529

**Pull request merged**